include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1233 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_MPI_HPP_
+#define GKO_PUBLIC_CORE_BASE_MPI_HPP_
+
+
+#include <map>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <type_traits>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+#if GKO_HAVE_MPI
+
+
+#include <mpi.h>
+
+
+namespace gko {
+namespace mpi {
+
+
+/*
+ * This enum specifies the threading type to be used when creating an MPI
+ * environment.
+ */
+enum class thread_type {
+    serialized = MPI_THREAD_SERIALIZED,
+    funneled = MPI_THREAD_FUNNELED,
+    single = MPI_THREAD_SINGLE,
+    multiple = MPI_THREAD_MULTIPLE
+};
+
+
+namespace detail {
+
+
+template <typename T>
+constexpr MPI_Datatype mpi_type_impl()
+{
+    return MPI_C_BOOL;
```

Comment:
I think both have issues, since they cause MPI to assume the wrong sizeof(T). I would suggest we use a specialized `struct mpi_type_impl{};` with a `static get_type()` function that is only declared for types that have a MPI counterpart. That way, we get a compiler error when attempting to communicate with types that MPI doesn't know. Users can always specialize them with their own MPI_Type if they need them (Maybe we need a RAII type with conditional ownership here as well?)